File: App/URLs/AwfulURLRouter.m
//  AwfulURLRouter.m
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "AwfulURLRouter.h"
#import "AwfulForumsClient.h"
#import "EmptyViewController.h"
#import <JLRoutes/JLRoutes.h>
#import <MRProgress/MRProgressOverlayView.h>
#import "PostsPageViewController.h"
#import "RapSheetViewController.h"
#import "SettingsViewController.h"
#import "Awful-Swift.h"

@implementation AwfulURLRouter
{
    JLRoutes *_routes;
}

- (id)initWithRootViewController:(UIViewController *)rootViewController managedObjectContext:(NSManagedObjectContext *)managedObjectContext
{
    if ((self = [super init])) {
        _rootViewController = rootViewController;
        _managedObjectContext = managedObjectContext;
    }
    return self;
}

- (instancetype)init
{
    NSAssert(nil, @"Use -initWithRootViewController:managedObjectContext: instead");
    return [self initWithRootViewController:nil managedObjectContext:nil];
}

- (JLRoutes *)routes
{
    if (_routes) return _routes;
    _routes = [JLRoutes new];
    __weak __typeof__(self) weakSelf = self;
    
    #pragma mark /forums/:forumID
    [_routes addRoute:@"/forums/:forumID" handler:^(NSDictionary *parameters) {
        __typeof__(self) self = weakSelf;
        ForumKey *key = [[ForumKey alloc] initWithForumID:parameters[@"forumID"]];
        Forum *forum = [Forum existingObjectForKey:key inManagedObjectContext:self.managedObjectContext];
        if (forum) {
            return [self jumpToForum:forum];
        } else {
            return NO;
        }
    }];
    
    #pragma mark /forums
    [_routes addRoute:@"/forums" handler:^BOOL(NSDictionary *parameters) {
        return !![weakSelf selectTopmostViewControllerContainingViewControllerOfClass:[ForumsTableViewController class]];
    }];
    
    #pragma mark /threads/:threadID/pages/:page
    [_routes addRoute:@"/threads/:threadID/pages/:page" handler:^(NSDictionary *parameters) {
        return [weakSelf showThreadWithParameters:parameters];
    }];
    
    #pragma mark /threads/:threadID
    [_routes addRoute:@"/threads/:threadID" handler:^(NSDictionary *parameters) {
        return [weakSelf showThreadWithParameters:parameters];
    }];
    
    #pragma mark /posts/:postID
    [_routes addRoute:@"/posts/:postID" handler:^(NSDictionary *parameters) {
        __typeof__(self) self = weakSelf;
        PostKey *key = [[PostKey alloc] initWithPostID:parameters[@"postID"]];
        Post *post = [Post existingObjectForKey:key inManagedObjectContext:self.managedObjectContext];
        if (post && post.page > 0) {
            PostsPageViewController *postsViewController = [[PostsPageViewController alloc] initWithThread:post.thread];
            [postsViewController loadPage:post.page updatingCache:YES updatingLastReadPost:YES];
            [postsViewController scrollPostToVisible:post];
            return [self showPostsViewController:postsViewController];
        }
        
        MRProgressOverlayView *overlay = [MRProgressOverlayView showOverlayAddedTo:self.rootViewController.view
                                                                             title:@"Locating Post"
                                                                              mode:MRProgressOverlayViewModeIndeterminate
                                                                          animated:YES];
        overlay.tintColor = [Theme currentTheme][@"tintColor"];
        __weak __typeof__(self) weakSelf = self;
        [[AwfulForumsClient client] locatePostWithID:key.postID andThen:^(NSError *error, Post *post, AwfulThreadPage page) {
            __typeof__(self) self = weakSelf;
            if (error) {
                overlay.titleLabelText = @"Post Not Found";
                overlay.mode = MRProgressOverlayViewModeCross;
                dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.7 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                    [overlay dismiss:YES];
                });
            } else {
                [overlay dismiss:YES completion:^{
                    PostsPageViewController *postsViewController = [[PostsPageViewController alloc] initWithThread:post.thread];
                    [postsViewController loadPage:page updatingCache:YES updatingLastReadPost:YES];
                    [postsViewController scrollPostToVisible:post];
                    [self showPostsViewController:postsViewController];
                    NSError *error;
                    if (![self.managedObjectContext save:&error]) {
                        NSLog(@"%s error saving managed object context: %@", __PRETTY_FUNCTION__, error);
                    }
                }];
            }
        }];
        return YES;
    }];
    
    #pragma mark /messages/:messageID
    [_routes addRoute:@"/messages/:messageID" handler:^(NSDictionary *parameters) {
        __typeof__(self) self = weakSelf;
        MessageListViewController *inbox = [self selectTopmostViewControllerContainingViewControllerOfClass:[MessageListViewController class]];
        if (!inbox) {
            return NO;
        }
        [inbox.navigationController popToViewController:inbox animated:NO];
        
        PrivateMessageKey *messageKey = [[PrivateMessageKey alloc] initWithMessageID:parameters[@"messageID"]];
        PrivateMessage *message = [PrivateMessage objectForKey:messageKey inManagedObjectContext:self.managedObjectContext];
        if (message) {
            [inbox showMessage:message];
            return YES;
        }
        
        MRProgressOverlayView *overlay = [MRProgressOverlayView showOverlayAddedTo:self.rootViewController.view
                                                                             title:@"Locating Message"
                                                                              mode:MRProgressOverlayViewModeIndeterminate
                                                                          animated:YES];
        overlay.tintColor = [Theme currentTheme][@"tintColor"];
        [[AwfulForumsClient sharedClient] readPrivateMessageWithKey:messageKey andThen:^(NSError *error, PrivateMessage *message) {
            if (error) {
                overlay.titleLabelText = @"Message Not Found";
                overlay.mode = MRProgressOverlayViewModeCross;
                dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.7 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                    [overlay dismiss:YES];
                });
            } else {
                [overlay dismiss:YES completion:^{
                    [inbox showMessage:message];
                }];
            }
        }];
        return YES;
    }];
    
    #pragma mark /messages
    [_routes addRoute:@"/messages" handler:^BOOL(NSDictionary *parameters) {
        return !![weakSelf selectTopmostViewControllerContainingViewControllerOfClass:[MessageListViewController class]];
    }];
    
    #pragma mark /bookmarks
    [_routes addRoute:@"/bookmarks" handler:^BOOL(NSDictionary *parameters) {
        return !![weakSelf selectTopmostViewControllerContainingViewControllerOfClass:[BookmarksTableViewController class]];
    }];
    
    #pragma mark /settings
    [_routes addRoute:@"/settings" handler:^BOOL(NSDictionary *parameters) {
        return !![weakSelf selectTopmostViewControllerContainingViewControllerOfClass:[SettingsViewController class]];
    }];
    
    #pragma mark /users/:userID
    [_routes addRoute:@"/users/:userID" handler:^(NSDictionary *parameters) {
        __typeof__(self) self = weakSelf;
        void (^success)() = ^(User *user) {
            ProfileViewController *profile = [[ProfileViewController alloc] initWithUser:user];
            [self.rootViewController presentViewController:[profile enclosingNavigationController] animated:YES completion:nil];
        };
        UserKey *key = [[UserKey alloc] initWithUserID:parameters[@"userID"] username:nil];
        User *user = [User existingObjectForKey:key inManagedObjectContext:self.managedObjectContext];
        if (user) {
            success(user);
            return YES;
        }
        [[AwfulForumsClient client] profileUserWithID:key.userID username:nil andThen:^(NSError *error, Profile *profile) {
            if (profile) {
                success(profile.user);
            } else if (error) {
                [self.rootViewController presentViewController:[UIAlertController alertWithTitle:@"Could Not Find User" error:error] animated:YES completion:nil];
            }
        }];
        return YES;
    }];
	
	#pragma mark /banlist/:userID
	[_routes addRoute:@"/banlist/:userID" handler:^BOOL(NSDictionary *parameters) {
		__typeof__(self) self = weakSelf;
		
		void (^success)() = ^(User *user) {
			RapSheetViewController *rapSheet = [[RapSheetViewController alloc] initWithUser:user];
			[self.rootViewController presentViewController:[rapSheet enclosingNavigationController] animated:YES completion:nil];
		};
        UserKey *key = [[UserKey alloc] initWithUserID:parameters[@"userID"] username:nil];
        User *user = [User existingObjectForKey:key inManagedObjectContext:self.managedObjectContext];
		if (user) {
			success(user);
			return YES;
		}
		
		[[AwfulForumsClient client] profileUserWithID:key.userID username:nil andThen:^(NSError *error, Profile *profile) {
			if (profile) {
				success(profile.user);
			} else if (error) {
                [self.rootViewController presentViewController:[UIAlertController alertWithTitle:@"Could Not Find User" error:error] animated:YES completion:nil];
			}
		}];
		return YES;
	}];
	
    #pragma mark /banlist
	[_routes addRoute:@"/banlist" handler:^BOOL(NSDictionary *parameters) {
		__typeof__(self) self = weakSelf;
		
		RapSheetViewController *rapSheet = [[RapSheetViewController alloc] initWithUser:nil];
		[self.rootViewController presentViewController:[rapSheet enclosingNavigationController] animated:YES completion:nil];
		
		return YES;
	}];
    
    return _routes;
}

- (void)doneWithProfile
{
    [self.rootViewController dismissViewControllerAnimated:YES completion:nil];
}

- (BOOL)jumpToForum:(Forum *)forum
{
    ThreadsTableViewController *threadList = [self.rootViewController awful_firstDescendantViewControllerOfClass:[ThreadsTableViewController class]];
    if ([threadList.forum isEqual:forum]) {
        [threadList.navigationController popToViewController:threadList animated:YES];
        return !![self selectTopmostViewControllerContainingViewControllerOfClass:threadList.class];
    } else {
        ForumsTableViewController *forumsList = [self.rootViewController awful_firstDescendantViewControllerOfClass:[ForumsTableViewController class]];
        [forumsList.navigationController popToViewController:forumsList animated:NO];
        [forumsList openForum:forum animated:NO];
        return !![self selectTopmostViewControllerContainingViewControllerOfClass:forumsList.class];
    }
}

- (id)selectTopmostViewControllerContainingViewControllerOfClass:(Class)class
{
    UISplitViewController *splitViewController = (UISplitViewController *)self.rootViewController;
    UITabBarController *tabBarController = splitViewController.viewControllers.firstObject;
    for (UIViewController *topmost in tabBarController.viewControllers) {
        UIViewController *match = [topmost awful_firstDescendantViewControllerOfClass:class];
        if (match) {
            tabBarController.selectedViewController = topmost;
            [splitViewController awful_showPrimaryViewController];
            return match;
        }
    }
    return nil;
}

- (BOOL)showThreadWithParameters:(NSDictionary *)parameters
{
    // TODO don't unilaterally create a thread based on the ID. If we don't already know of it, try to fetch it.
    ThreadKey *threadKey = [[ThreadKey alloc] initWithThreadID:parameters[@"threadID"]];
    Thread *thread = [Thread objectForKey:threadKey inManagedObjectContext:self.managedObjectContext];
    PostsPageViewController *postsViewController;
    NSString *userID = parameters[@"userid"];
    if (userID.length > 0) {
        UserKey *userKey = [[UserKey alloc] initWithUserID:userID username:nil];
        User *user = [User objectForKey:userKey inManagedObjectContext:self.managedObjectContext];
        postsViewController = [[PostsPageViewController alloc] initWithThread:thread author:user];
    } else {
        postsViewController = [[PostsPageViewController alloc] initWithThread:thread];
    }
    NSError *error;
    if (![self.managedObjectContext save:&error]) {
        NSLog(@"%s error saving managed object context: %@", __PRETTY_FUNCTION__, error);
    }
    NSString *pageString = parameters[@"page"];
    AwfulThreadPage page = AwfulThreadPageNone;
    if (userID.length == 0) {
        if ([pageString isEqualToString:@"last"]) {
            page = AwfulThreadPageLast;
        } else if ([pageString isEqualToString:@"unread"]) {
            page = AwfulThreadPageNextUnread;
        }
    }
    if (page == AwfulThreadPageNone) {
        page = pageString.integerValue ?: thread.beenSeen ? AwfulThreadPageNextUnread : 1;
    }
    [postsViewController loadPage:page updatingCache:YES updatingLastReadPost:YES];
    if (parameters[@"post"]) {
        PostKey *postKey = [[PostKey alloc] initWithPostID:parameters[@"post"]];
        [postsViewController scrollPostToVisible:[Post objectForKey:postKey inManagedObjectContext:self.managedObjectContext]];
    }
    return [self showPostsViewController:postsViewController];
}

- (BOOL)showPostsViewController:(PostsPageViewController *)postsViewController
{
    postsViewController.restorationIdentifier = @"Posts from URL";
    
    // Showing a posts view controller as a result of opening a URL is not the same as simply showing a detail view controller. We want to push it on to an existing navigation stack. Which one depends on how the split view is currently configured.
    UINavigationController *targetNavigationController;
    UISplitViewController *splitViewController = (UISplitViewController *)self.rootViewController;
    if (splitViewController.viewControllers.count == 2) {
        targetNavigationController = splitViewController.viewControllers[1];
    } else {
        UITabBarController *tabBarController = splitViewController.viewControllers[0];
        targetNavigationController = (UINavigationController *)tabBarController.selectedViewController;
    }
    
    // If the detail view controller is empty, showing the posts view controller actually is as simple as showing a detail view controller, and we can exit early.
    if ([targetNavigationController.topViewController isKindOfClass:[EmptyViewController class]]) {
        [splitViewController showDetailViewController:postsViewController sender:self];
        return YES;
    }
    
    // Posts view controllers by default hide the bottom bar when pushed. This moves the tab bar controller's tab bar out of the way, making room for the toolbar. However, if some earlier posts view controller has already done this for us, and we went ahead oblivious, we would hide our own toolbar!
    if ([targetNavigationController.topViewController isKindOfClass:[PostsPageViewController class]]) {
        postsViewController.hidesBottomBarWhenPushed = NO;
    }
    
    [targetNavigationController pushViewController:postsViewController animated:YES];
    return YES;
}

- (BOOL)routeURL:(NSURL *)URL
{
    if ([URL.scheme caseInsensitiveCompare:@"awful"] != NSOrderedSame) return NO;
    return [self.routes routeURL:URL];
}

@end

File: App/URLs/AwfulURLRouter.swift
//  AwfulURLRouter.swift
//
//  Copyright 2016 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import AwfulCore
import AwfulSettings
import AwfulTheming
import CoreData
import MRProgress
import UIKit

/// Translates URLs with the scheme "awful" into an appropriate shown screen.
struct AwfulURLRouter {

    @FoilDefaultStorage(Settings.enableHaptics) private var enableHaptics
    private let managedObjectContext: NSManagedObjectContext
    private let rootViewController: UIViewController?
    weak var coordinator: (any MainCoordinator)?
    
    /**
        - parameter rootViewController: The application's root view controller (optional for SwiftUI-only mode).
        - parameter managedObjectContext: The managed object context we use to find forums, threads, and posts.
     */
    init(rootViewController: UIViewController?, managedObjectContext: NSManagedObjectContext) {
        self.rootViewController = rootViewController
        self.managedObjectContext = managedObjectContext
    }
    
    /// Show the screen appropriate for an "awful" URL.
    @discardableResult
    func route(_ route: AwfulRoute) -> Bool {
        
        if enableHaptics {
            UIImpactFeedbackGenerator(style: .medium).impactOccurred()
        }
        
        switch route {
        case .bookmarks:
            if let coordinator = coordinator {
                coordinator.navigateToTab(.bookmarks)
                return true
            }
            return selectTopmostViewController(containingViewControllerOfClass: BookmarksTableViewController.self) != nil

        case let .forum(id: forumID):
            if let coordinator = coordinator {
                return coordinator.navigateToForumWithID(forumID)
            }
            let key = ForumKey(forumID: forumID)
            guard let forum = Forum.existingObjectForKey(objectKey: key, in: managedObjectContext) else { return false }
            return jumpToForum(forum)

        case .forumList:
            if let coordinator = coordinator {
                coordinator.navigateToTab(.forums)
                return true
            }
            return selectTopmostViewController(containingViewControllerOfClass: ForumsTableViewController.self) != nil

        case .lepersColony:
            if let coordinator = coordinator {
                coordinator.navigateToTab(.lepers)
                return true
            }
            guard let rootViewController = rootViewController else { return false }
            let rapSheetVC = RapSheetViewController(user: nil)
            rootViewController.present(rapSheetVC.enclosingNavigationController, animated: true)
            return true

        case let .message(id: messageID):
            if let coordinator = coordinator {
                return coordinator.navigateToMessageWithID(messageID)
            }
            guard let inbox = selectTopmostViewController(containingViewControllerOfClass: MessageListViewController.self) else { return false }
            _ = inbox.navigationController?.popToViewController(inbox, animated: false)

            let key = PrivateMessageKey(messageID: messageID)
            if let message = PrivateMessage.existingObjectForKey(objectKey: key, in: managedObjectContext) {
                inbox.showMessage(message)
                return true
            }

            guard let rootViewController = rootViewController,
                  let rootView = rootViewController.view else { return false }
            let overlay = MRProgressOverlayView.showOverlayAdded(to: rootView, title: "Locating Message", mode: .indeterminate, animated: true)!
            overlay.tintColor = Theme.defaultTheme()["tintColor"]

            Task { @MainActor in
                do {
                    let message = try await ForumsClient.shared.readPrivateMessage(identifiedBy: key)
                    overlay.dismiss(true, completion: {
                        inbox.showMessage(message)
                    })
                } catch {
                    overlay.titleLabelText = "Message Not Found"
                    overlay.mode = .cross
                    try? await Task.sleep(timeInterval: 0.7)
                    overlay.dismiss(true)
                }
            }
            return true

        case .messagesList:
            if let coordinator = coordinator {
                coordinator.navigateToTab(.messages)
                return true
            }
            return selectTopmostViewController(containingViewControllerOfClass: MessageListViewController.self) != nil

        case let .post(id: postID, updateSeen):
            print("🔗 AwfulURLRouter: Handling .post route with postID: \(postID)")
            if let coordinator = coordinator {
                print("🔗 AwfulURLRouter: Calling coordinator.navigateToPostWithID(\(postID))")
                let result = coordinator.navigateToPostWithID(postID)
                print("🔗 AwfulURLRouter: coordinator.navigateToPostWithID returned: \(result)")
                return result
            }
            let key = PostKey(postID: postID)

            var updateLastRead: Bool {
                switch updateSeen {
                case .noseen: return false
                case .seen: return true
                }
            }
            
            if let post = Post.existingObjectForKey(objectKey: key, in: managedObjectContext),
               let thread = post.thread,
               post.page > 0
            {
                let postsVC = PostsPageViewController(thread: thread)
                postsVC.jumpToPostIDAfterLoading = post.postID
                postsVC.loadPage(.specific(post.page), updatingCache: true, updatingLastReadPost: updateLastRead)
                return showPostsViewController(postsVC)
            }

            guard let rootViewController = rootViewController,
                  let rootView = rootViewController.view else { return false }
            let overlay = MRProgressOverlayView.showOverlayAdded(to: rootView, title: "Locating Post", mode: .indeterminate, animated: true)!
            overlay.tintColor = Theme.defaultTheme()["tintColor"]

            Task { @MainActor in
                do {
                    let (post, page) = try await ForumsClient.shared.locatePost(id: key.postID, updateLastReadPost: updateLastRead)
                    overlay.dismiss(true) {
                        guard let thread = post.thread else { return }

                        if let coordinator = self.coordinator {
                            // TODO: This doesn't scroll to the post.
                            coordinator.navigateToThread(thread, page: page)
                        } else {
                            let postsVC = PostsPageViewController(thread: thread)
                            postsVC.jumpToPostIDAfterLoading = post.postID
                            postsVC.loadPage(page, updatingCache: true, updatingLastReadPost: true)
                            _ = self.showPostsViewController(postsVC)
                        }
                    }
                } catch {
                    overlay.titleLabelText = "Post Not Found"
                    overlay.mode = .cross
                    try? await Task.sleep(timeInterval: 3)
                    overlay.dismiss(true)
                }
            }
            return true

        case let .profile(userID: userID):
            if let coordinator = coordinator {
                coordinator.presentUserProfile(userID: userID)
                return true
            }
            guard let rootViewController = rootViewController else { return false }
            Task { @MainActor in
                do {
                    let user = try await fetchUser(withUserID: userID)
                    let profileVC = ProfileViewController(user: user)
                    rootViewController.present(profileVC.enclosingNavigationController, animated: true)
                } catch {
                    let alert = UIAlertController(title: "Could Not Find User", error: error)
                    rootViewController.present(alert, animated: true)
                }
            }
            return true

        case let .rapSheet(userID: userID):
            if let coordinator = coordinator {
                coordinator.presentRapSheet(userID: userID)
                return true
            }
            guard let rootViewController = rootViewController else { return false }
            Task { @MainActor in
                do {
                    let user = try await fetchUser(withUserID: userID)
                    let rapSheetVC = RapSheetViewController(user: user)
                    rootViewController.present(rapSheetVC.enclosingNavigationController, animated: true)
                } catch {
                    let alert = UIAlertController(title: "Could Not Find User", error: error)
                    rootViewController.present(alert, animated: true)
                }
            }
            return true

        case .settings:
            if let coordinator = coordinator {
                coordinator.navigateToTab(.settings)
                return true
            }
            return selectTopmostViewController(containingViewControllerOfClass: SettingsViewController.self) != nil

        case let .threadPage(threadID: threadID, page: page, updateSeen):
            if let coordinator = coordinator {
                return coordinator.navigateToThreadWithID(threadID, page: page, author: nil)
            }
            return showThread(threadID, page: page, updateSeen: updateSeen)

        case let .threadPageSingleUser(threadID: threadID, userID: userID, page: page, updateSeen):
            if let coordinator = coordinator {
                // Find the author user
                let userKey = UserKey(userID: userID, username: nil)
                let author = User.existingObjectForKey(objectKey: userKey, in: managedObjectContext)
                return coordinator.navigateToThreadWithID(threadID, page: page, author: author)
            }
            return showThread(threadID, page: page, justPostsByUser: userID, updateSeen: updateSeen)
        }
    }
    
    private func jumpToForum(_ forum: Forum) -> Bool {
        if enableHaptics {
            UIImpactFeedbackGenerator(style: .medium).impactOccurred()
        }
        guard let rootViewController = rootViewController else { return false }
        
        if let threadsVC = rootViewController.firstDescendant(ofType: ThreadsTableViewController.self),
           threadsVC.forum === forum
        {
            _ = threadsVC.navigationController?.popToViewController(threadsVC, animated: true)
            return selectTopmostViewController(containingViewControllerOfClass: ThreadsTableViewController.self) != nil
        }
        
        if let forumsVC = rootViewController.firstDescendant(ofType: ForumsTableViewController.self) {
            _ = forumsVC.navigationController?.popToViewController(forumsVC, animated: false)
            forumsVC.openForum(forum, animated: false)
            return selectTopmostViewController(containingViewControllerOfClass: ForumsTableViewController.self) != nil
        }
        
        return false
    }
    
    private func selectTopmostViewController<VC: UIViewController>(
        containingViewControllerOfClass klass: VC.Type
    ) -> VC? {
        if enableHaptics {
            UIImpactFeedbackGenerator(style: .medium).impactOccurred()
        }
        guard let rootViewController = rootViewController,
              let splitVC = rootViewController.children.first as? UISplitViewController,
              let tabBarVC = splitVC.viewControllers.first as? UITabBarController
              else { return nil }
        for topmost in tabBarVC.viewControllers ?? [] {
            guard let match = topmost.firstDescendant(ofType: VC.self) else { continue }
            tabBarVC.selectedViewController = topmost
            splitVC.showPrimaryViewController()
            return match
        }
        return nil
    }
    
    private func showThread(
        _ threadID: String,
        page: ThreadPage,
        justPostsByUser userID: String? = nil,
        updateSeen: AwfulRoute.UpdateSeen
    ) -> Bool {
        if let coordinator = coordinator {
            let threadKey = ThreadKey(threadID: threadID)
            let thread = AwfulThread.objectForKey(objectKey: threadKey, in: managedObjectContext)
            
            var author: User?
            if let userID = userID, !userID.isEmpty {
                let userKey = UserKey(userID: userID, username: nil)
                author = User.objectForKey(objectKey: userKey, in: managedObjectContext)
            }
            
            // This doesn't handle updateSeen yet.
            coordinator.navigateToThread(thread, page: page, author: author)
            
            return true
        }
        
        let threadKey = ThreadKey(threadID: threadID)
        let thread = AwfulThread.objectForKey(objectKey: threadKey, in: managedObjectContext) 
        let postsVC: PostsPageViewController
        if let userID = userID, !userID.isEmpty {
            let userKey = UserKey(userID: userID, username: nil)
            let user = User.objectForKey(objectKey: userKey, in: managedObjectContext) 
            postsVC = PostsPageViewController(thread: thread, author: user)
        } else {
            postsVC = PostsPageViewController(thread: thread)
        }
        
        try! managedObjectContext.save()

        var updateLastRead: Bool {
            switch updateSeen {
            case .noseen: return false
            case .seen: return true
            }
        }
        postsVC.loadPage(page, updatingCache: true, updatingLastReadPost: updateLastRead)
        
        return showPostsViewController(postsVC)
    }
    
    private func showPostsViewController(_ postsVC: PostsPageViewController) -> Bool {
        postsVC.restorationIdentifier = "Posts from URL"
        
        // Showing a posts view controller as a result of opening a URL is not the same as simply showing a detail view controller. We want to push it on to an existing navigation stack. Which one depends on how the split view is currently configured.
        let targetNav: UINavigationController
        guard let rootViewController = rootViewController,
              let splitVC = rootViewController.children.first as? UISplitViewController else { return false }
        if splitVC.viewControllers.count == 2 {
            targetNav = splitVC.viewControllers[1] as! UINavigationController
        } else {
            guard let tabBarVC = splitVC.viewControllers[0] as? UITabBarController else { return false }
            targetNav = tabBarVC.selectedViewController as! UINavigationController
        }
        
        // If the detail view controller is empty, showing the posts view controller actually is as simple as showing a detail view controller, and we can exit early.
        if targetNav.topViewController is EmptyViewController {
            splitVC.showDetailViewController(postsVC, sender: self)
            return true
        }
        
        // Posts view controllers by default hide the bottom bar when pushed. This moves the tab bar controller's tab bar out of the way, making room for the toolbar. However, if some earlier posts view controller has already done this for us, and we went ahead oblivious, we would hide our own toolbar!
        if targetNav.topViewController is PostsPageViewController {
            postsVC.hidesBottomBarWhenPushed = false
        }
        
        targetNav.pushViewController(postsVC, animated: true)
        return true
    }
    
    private func fetchUser(withUserID userID: String) async throws -> User {
        let key = UserKey(userID: userID, username: nil)
        if let user = User.existingObjectForKey(objectKey: key, in: managedObjectContext) {
            return user
        }
        
        return try await ForumsClient.shared.profileUser(.userID(userID)).user
    }
}
